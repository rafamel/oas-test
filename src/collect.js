export default function collect(oas) {
  const collection = {};
  Object.entries(oas.paths).forEach(([path, inner]) => {
    Object.entries(inner).forEach(([method, operation]) => {
      if (!operation.operationId) return;
      collection[operation.operationId] = {
        path,
        method,
        operation
      };
    });
  });
  return collection;
}
