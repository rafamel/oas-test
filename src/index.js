export { default } from './TestRunner';
export * from './helpers';
