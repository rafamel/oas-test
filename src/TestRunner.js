import deep from 'lodash.clonedeep';
// sync alternative: json-ref-lite
import $RefParser from 'json-schema-ref-parser';
import SwaggerParser from 'swagger-parser';
import chalk from 'chalk';
import collect from './collect';

export const STORE_SYMBOL = Symbol('store');

export default class TestRunner {
  constructor() {
    this._first = false;
    this[STORE_SYMBOL] = {
      app: null,
      url: null,
      settings: {
        assert: (bool) => expect(bool).toBe(true),
        describe: describe,
        test: test,
        logger: console,
        logall: false
      },
      hooks: {
        data: [],
        request: [],
        response: []
      },
      context: {},
      oas: Promise.resolve(null),
      collection: Promise.resolve(null),
      assertlog: (condition, description) => {
        const { settings } = this[STORE_SYMBOL];
        if (
          settings.logger &&
          !condition &&
          (settings.logall || !this._first)
        ) {
          this._first = true;
          settings.logger.log(chalk.red(description));
        }
        settings.assert(condition);
      }
    };
  }
  get context() {
    return this[STORE_SYMBOL].context;
  }
  load({ app, url, oas } = {}) {
    this[STORE_SYMBOL].app = app || null;
    this[STORE_SYMBOL].url = url || null;

    if (oas) {
      // Dereference and validate oas
      const promise = Promise.resolve(oas)
        .then((oas) => $RefParser.dereference(deep(oas)))
        .then((oas) => SwaggerParser.validate(oas))
        .catch((e) => {
          throw e;
        });
      this[STORE_SYMBOL].oas = promise;
      this[STORE_SYMBOL].collection = promise.then(collect);
    } else {
      this[STORE_SYMBOL].oas = Promise.resolve(null);
      this[STORE_SYMBOL].collection = Promise.resolve(null);
    }

    return this;
  }
  settings(obj = {}) {
    Object.assign(this[STORE_SYMBOL].settings, obj);
    return this;
  }
  hook(type, cb) {
    const hookTypes = ['data', 'request', 'response'];
    if (!hookTypes.includes(type)) {
      throw Error(`Hooks can only be of types ${hookTypes.join(', ')}.`);
    }
    this[STORE_SYMBOL].hooks[type].push(cb);
    return this;
  }
  clone() {
    const testRunner = new TestRunner();
    testRunner[STORE_SYMBOL] = deep(this[STORE_SYMBOL]);
    return testRunner;
  }
  self(cb) {
    // eslint-disable-next-line standard/no-callback-literal
    if (cb) cb(this);
    return this;
  }
  test(tests = {}) {
    const store = this[STORE_SYMBOL];

    if (!store.app && !store.url) {
      throw Error(
        "TestRunner.load() hasn't been called with a valid app, url, or both"
      );
    }

    return this;
  }
}
