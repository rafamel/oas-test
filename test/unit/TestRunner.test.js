import TestRunner from '../../src';
import { STORE_SYMBOL } from '../../src/TestRunner';
import oas from '../setup/static-oas.json';
import deep from 'lodash.clonedeep';

describe(`Initial state`, () => {
  test(`oas & collection are null resolving promises`, async (done) => {
    const tr = new TestRunner();

    expect(tr[STORE_SYMBOL].oas).toBeInstanceOf(Promise);
    expect(tr[STORE_SYMBOL].collection).toBeInstanceOf(Promise);
    expect(await tr[STORE_SYMBOL].oas).toBe(null);
    expect(await tr[STORE_SYMBOL].collection).toBe(null);
    done();
  });
  test(`default settings.assert should assert`, () => {
    expect.assertions(1);
    const tr = new TestRunner();
    const assert = tr[STORE_SYMBOL].settings.assert;

    assert(true);
  });
  test(`default settings.errorHandler throws`, () => {
    const tr = new TestRunner();
    const errorHandler = tr[STORE_SYMBOL].settings.errorHandler;

    expect(() => errorHandler(Error())).toThrowError();
  });
  test(`hooks for data, response, request, are empty arrays`, () => {
    const tr = new TestRunner();

    expect(typeof tr[STORE_SYMBOL].hooks).toBe('object');
    expect(tr[STORE_SYMBOL].hooks.data).toEqual([]);
    expect(tr[STORE_SYMBOL].hooks.request).toEqual([]);
    expect(tr[STORE_SYMBOL].hooks.response).toEqual([]);
  });
});

describe(`tr.context`, () => {
  test(`tr.context is stored object`, () => {
    const tr = new TestRunner();
    expect(typeof tr.context).toBe('object');
    expect(tr.context).toBe(tr[STORE_SYMBOL].context);
  });
});

describe(`tr.load()`, () => {
  test(`loads all properties; sets collection; dereferences oas`, async (done) => {
    const tr = new TestRunner();
    const app = {};

    expect(() => tr.load({ app, url: 'url', oas })).not.toThrow();
    expect(tr[STORE_SYMBOL].app).toBe(app);
    expect(tr[STORE_SYMBOL].url).toBe('url');
    expect(tr[STORE_SYMBOL].oas).toBeInstanceOf(Promise);
    expect(tr[STORE_SYMBOL].collection).toBeInstanceOf(Promise);

    const trOas = await tr[STORE_SYMBOL].oas;
    const trCollection = await tr[STORE_SYMBOL].collection;

    expect(trOas).toMatchSnapshot();
    expect(trCollection).toMatchSnapshot();
    done();
  });
  test(`doesn't fail when no args are passed`, () => {
    const tr = new TestRunner();

    expect(() => tr.load()).not.toThrow();
  });
  test(`returns instance`, () => {
    const tr = new TestRunner();

    expect(tr.load()).toBe(tr);
    expect(tr.load({ app: {}, oas, url: 'url' })).toBe(tr);
  });
  test(`adding an OAS updates tr.promise`, async () => {
    expect.assertions(2);
    const tr = new TestRunner();

    const p1 = tr.promise;
    let time1, time2;
    tr.load({ url: 'url', oas });
    tr[STORE_SYMBOL].oas.then(() => (time1 = Date.now()));
    tr.promise.then(() => (time2 = Date.now()));

    await tr.promise;

    expect(tr.promise).not.toBe(p1);
    expect(time2).toBeGreaterThanOrEqual(time1);
  });
  test(`throws error (errorHandler) when oas is not valid`, async () => {
    expect.assertions(1);
    let err;
    const tr = new TestRunner().settings({ errorHandler: (e) => (err = e) });

    tr.load({ oas: { ...oas, invalidProp: true }, url: 'url' });

    await tr.promise.catch(() => {});
    expect(err).toBeInstanceOf(Error);
  });
  test(`overwrites values for null, preserves for undefined`, () => {
    const tr = new TestRunner().load({ app: 1, url: 2, oas });
    const pOas = tr[STORE_SYMBOL].oas;

    tr.load({});
    expect(tr[STORE_SYMBOL].app).toBe(1);
    expect(tr[STORE_SYMBOL].url).toBe(2);
    expect(tr[STORE_SYMBOL].oas).toBe(pOas);

    tr.load({ app: 3 });
    expect(tr[STORE_SYMBOL].app).toBe(3);
    expect(tr[STORE_SYMBOL].url).toBe(2);
    expect(tr[STORE_SYMBOL].oas).toBe(pOas);

    tr.load({ url: 4 });
    expect(tr[STORE_SYMBOL].app).toBe(3);
    expect(tr[STORE_SYMBOL].url).toBe(4);
    expect(tr[STORE_SYMBOL].oas).toBe(pOas);

    tr.load({ oas });
    expect(tr[STORE_SYMBOL].app).toBe(3);
    expect(tr[STORE_SYMBOL].url).toBe(4);
    expect(tr[STORE_SYMBOL].oas).not.toBe(pOas);
    expect(tr[STORE_SYMBOL].oas).toBeInstanceOf(Promise);
  });
});

describe(`tr.settings()`, () => {
  test(`merges settings`, () => {
    const tr = new TestRunner();
    const settings = deep(tr[STORE_SYMBOL].settings);
    const added = { assert: 1, describe: 2, test: 3, logger: 4 };
    tr.settings(added);
    expect(tr[STORE_SYMBOL].settings).toEqual({ ...settings, ...added });
  });
  test(`doesn't throw when no args are passed`, () => {
    const tr = new TestRunner();
    expect(() => tr.settings()).not.toThrow();
  });
  test(`returns instance`, () => {
    const tr = new TestRunner();
    expect(tr.settings()).toBe(tr);
    expect(tr.settings({})).toBe(tr);
  });
});

describe(`tr.hook()`, () => {
  test(`fails for invalid types, doesn't for valid types`, () => {
    const tr = new TestRunner();
    expect(() => tr.hook('response', () => {})).not.toThrow();
    expect(() => tr.hook('request', () => {})).not.toThrow();
    expect(() => tr.hook('data', () => {})).not.toThrow();
    expect(() => tr.hook('other', () => {})).toThrowError();
  });
  test(`returns instance`, () => {
    const tr = new TestRunner();
    expect(tr.hook('response', () => {})).toBe(tr);
    expect(tr.hook('request', () => {})).toBe(tr);
    expect(tr.hook('data', () => {})).toBe(tr);
  });
  test(`adds data hooks`, () => {
    const tr = new TestRunner();
    const a = () => {};
    const b = () => {};
    const c = () => {};

    tr.hook('data', a)
      .hook('data', b)
      .hook('data', c);

    expect(tr[STORE_SYMBOL].hooks.data).toEqual([a, b, c]);
  });
  test(`adds request hooks`, () => {
    const tr = new TestRunner();
    const a = () => {};
    const b = () => {};
    const c = () => {};

    tr.hook('request', a)
      .hook('request', b)
      .hook('request', c);

    expect(tr[STORE_SYMBOL].hooks.request).toEqual([a, b, c]);
  });
  test(`adds response hooks`, () => {
    const tr = new TestRunner();
    const a = () => {};
    const b = () => {};
    const c = () => {};

    tr.hook('response', a)
      .hook('response', b)
      .hook('response', c);

    expect(tr[STORE_SYMBOL].hooks.response).toEqual([a, b, c]);
  });
});

describe(`tr.clone()`, () => {
  test(`clones instance`, () => {
    const tr = new TestRunner();

    const a = () => {};
    const b = () => {};

    tr.load({ app: {}, url: '' })
      .hook('data', a)
      .hook('request', b);

    const cloned = tr.clone();

    expect(cloned).toBeInstanceOf(TestRunner);

    const trStore = { ...tr[STORE_SYMBOL], settings: {} };
    const clonedStore = { ...cloned[STORE_SYMBOL], settings: {} };

    expect({ ...trStore, assertlog: 1 }).toEqual({
      ...clonedStore,
      assertlog: 1
    });
    expect(trStore.assertlog).not.toBe(clonedStore.assertlog);
    expect(trStore.assertlog).toBeInstanceOf(Function);
    expect(clonedStore.assertlog).toBeInstanceOf(Function);

    tr.hook('response', () => {});
    expect(trStore).not.toEqual(clonedStore);
  });
});

describe(`tr.self()`, () => {
  test(`receives tr`, (done) => {
    const tr = new TestRunner();
    tr.self((tr2) => {
      expect(tr2).toBe(tr);
      done();
    });
  });
  test(`doesn't throw when no args are passed`, () => {
    const tr = new TestRunner();
    expect(() => tr.self()).not.toThrow();
  });
  test(`returns instance`, () => {
    const tr = new TestRunner();

    expect(tr.self()).toBe(tr);
    expect(tr.self(() => {})).toBe(tr);
  });
});

describe(`tr.test()`, () => {
  test(`doesn't throw when no args are passed`, () => {
    const tr = new TestRunner();
    expect(() => tr.load({ url: 'url' }).test()).not.toThrow();
    expect(() => tr.load({ url: 'url' }).test({})).not.toThrow();
  });
  test(`returns instance`, () => {
    const tr = new TestRunner();

    expect(tr.load({ url: 'url' }).test()).toBe(tr);
    expect(tr.load({ url: 'url' }).test({})).toBe(tr);
  });
  test(`throws when no url or app are passed`, () => {
    const tr = new TestRunner();

    expect(() => tr.load({ oas }).test()).toThrowError();
    expect(() => tr.load({}).test()).toThrowError();
    expect(() => tr.test()).toThrowError();
  });
  test(`doesn't throw when  url or app are passed`, () => {
    const tr = new TestRunner();

    expect(() => tr.load({ url: 'url' }).test()).not.toThrow();
    expect(() => tr.load({ app: {} }).test()).not.toThrow();
  });
});
