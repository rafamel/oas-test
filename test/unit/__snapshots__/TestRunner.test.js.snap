// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tr.load() loads all properties; sets collection; dereferences oas 1`] = `
Object {
  "components": Object {
    "callbacks": Object {},
    "examples": Object {},
    "headers": Object {},
    "links": Object {},
    "parameters": Object {
      "id": Object {
        "description": "ID of the item to return",
        "in": "path",
        "name": "id",
        "required": true,
        "schema": Object {
          "minimum": 0,
          "type": "integer",
        },
      },
    },
    "requestBodies": Object {
      "TodoCreate": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "done": Object {
                  "type": "boolean",
                },
                "name": Object {
                  "maxLength": 255,
                  "minLength": 1,
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "done",
              ],
              "type": "object",
            },
          },
        },
        "required": true,
      },
      "TodoPatch": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "minProperties": 1,
              "properties": Object {
                "done": Object {
                  "type": "boolean",
                },
                "name": Object {
                  "maxLength": 255,
                  "minLength": 1,
                  "type": "string",
                },
              },
              "type": "object",
            },
          },
        },
        "required": true,
      },
    },
    "responses": Object {
      "Todo": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "data": Object {
                  "properties": Object {
                    "done": Object {
                      "type": "boolean",
                    },
                    "id": Object {
                      "minimum": 0,
                      "type": "integer",
                    },
                    "name": Object {
                      "maxLength": 255,
                      "minLength": 1,
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "name",
                    "done",
                    "user_id",
                    "created_at",
                    "id",
                  ],
                  "type": "object",
                },
                "status": Object {
                  "enum": Array [
                    "success",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "status",
                "data",
              ],
              "type": "object",
            },
          },
        },
        "description": "Sucessful response",
      },
      "Todos": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "data": Object {
                  "items": Object {
                    "properties": Object {
                      "done": Object {
                        "type": "boolean",
                      },
                      "id": Object {
                        "minimum": 0,
                        "type": "integer",
                      },
                      "name": Object {
                        "maxLength": 255,
                        "minLength": 1,
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "name",
                      "done",
                      "user_id",
                      "created_at",
                      "id",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
                "status": Object {
                  "enum": Array [
                    "success",
                  ],
                  "type": "string",
                },
              },
              "required": Array [
                "status",
                "data",
              ],
              "type": "object",
            },
          },
        },
        "description": "Sucessful response",
      },
    },
    "schemas": Object {
      "Todo": Object {
        "properties": Object {
          "done": Object {
            "type": "boolean",
          },
          "id": Object {
            "minimum": 0,
            "type": "integer",
          },
          "name": Object {
            "maxLength": 255,
            "minLength": 1,
            "type": "string",
          },
        },
        "required": Array [
          "name",
          "done",
          "user_id",
          "created_at",
          "id",
        ],
        "type": "object",
      },
    },
    "securitySchemes": Object {
      "bearerAuth": Object {
        "bearerFormat": "JWT",
        "description": "User access token as header bearer token (Authorization: Bearer access_token)",
        "scheme": "bearer",
        "type": "http",
      },
    },
  },
  "info": Object {
    "description": "Powerful OpenAPI 3.0 based API testing, simplified. OAS 3 specification.",
    "license": Object {
      "name": "MIT",
    },
    "title": "Powerful OpenAPI 3.0 based API testing, simplified.",
    "version": "0.0.0",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/api/todo": Object {
      "get": Object {
        "operationId": "listTodo",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "data": Object {
                      "items": Object {
                        "properties": Object {
                          "done": Object {
                            "type": "boolean",
                          },
                          "id": Object {
                            "minimum": 0,
                            "type": "integer",
                          },
                          "name": Object {
                            "maxLength": 255,
                            "minLength": 1,
                            "type": "string",
                          },
                        },
                        "required": Array [
                          "name",
                          "done",
                          "user_id",
                          "created_at",
                          "id",
                        ],
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "enum": Array [
                        "success",
                      ],
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "status",
                    "data",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Sucessful response",
          },
        },
        "security": Array [
          Object {
            "bearerAuth": Array [],
          },
        ],
        "summary": "listTodo",
        "tags": Array [
          "todo",
        ],
      },
      "post": Object {
        "operationId": "createTodo",
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "done": Object {
                    "type": "boolean",
                  },
                  "name": Object {
                    "maxLength": 255,
                    "minLength": 1,
                    "type": "string",
                  },
                },
                "required": Array [
                  "name",
                  "done",
                ],
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "data": Object {
                      "properties": Object {
                        "done": Object {
                          "type": "boolean",
                        },
                        "id": Object {
                          "minimum": 0,
                          "type": "integer",
                        },
                        "name": Object {
                          "maxLength": 255,
                          "minLength": 1,
                          "type": "string",
                        },
                      },
                      "required": Array [
                        "name",
                        "done",
                        "user_id",
                        "created_at",
                        "id",
                      ],
                      "type": "object",
                    },
                    "status": Object {
                      "enum": Array [
                        "success",
                      ],
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "status",
                    "data",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Sucessful response",
          },
        },
        "security": Array [
          Object {
            "bearerAuth": Array [],
          },
        ],
        "summary": "createTodo",
        "tags": Array [
          "todo",
        ],
      },
    },
    "/api/todo/{id}": Object {
      "delete": Object {
        "operationId": "deleteTodo",
        "parameters": Array [
          Object {
            "description": "ID of the item to return",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "minimum": 0,
              "type": "integer",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "data": Object {
                      "enum": Array [
                        true,
                      ],
                      "type": "boolean",
                    },
                    "status": Object {
                      "enum": Array [
                        "success",
                      ],
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "status",
                    "data",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Successful operation",
          },
        },
        "security": Array [
          Object {
            "bearerAuth": Array [],
          },
        ],
        "summary": "deleteTodo",
        "tags": Array [
          "todo",
        ],
      },
      "get": Object {
        "operationId": "showTodo",
        "parameters": Array [
          Object {
            "description": "ID of the item to return",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "minimum": 0,
              "type": "integer",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "data": Object {
                      "properties": Object {
                        "done": Object {
                          "type": "boolean",
                        },
                        "id": Object {
                          "minimum": 0,
                          "type": "integer",
                        },
                        "name": Object {
                          "maxLength": 255,
                          "minLength": 1,
                          "type": "string",
                        },
                      },
                      "required": Array [
                        "name",
                        "done",
                        "user_id",
                        "created_at",
                        "id",
                      ],
                      "type": "object",
                    },
                    "status": Object {
                      "enum": Array [
                        "success",
                      ],
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "status",
                    "data",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Sucessful response",
          },
        },
        "security": Array [
          Object {
            "bearerAuth": Array [],
          },
        ],
        "summary": "showTodo",
        "tags": Array [
          "todo",
        ],
      },
      "patch": Object {
        "operationId": "patchTodo",
        "parameters": Array [
          Object {
            "description": "ID of the item to return",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "minimum": 0,
              "type": "integer",
            },
          },
        ],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "minProperties": 1,
                "properties": Object {
                  "done": Object {
                    "type": "boolean",
                  },
                  "name": Object {
                    "maxLength": 255,
                    "minLength": 1,
                    "type": "string",
                  },
                },
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "data": Object {
                      "properties": Object {
                        "done": Object {
                          "type": "boolean",
                        },
                        "id": Object {
                          "minimum": 0,
                          "type": "integer",
                        },
                        "name": Object {
                          "maxLength": 255,
                          "minLength": 1,
                          "type": "string",
                        },
                      },
                      "required": Array [
                        "name",
                        "done",
                        "user_id",
                        "created_at",
                        "id",
                      ],
                      "type": "object",
                    },
                    "status": Object {
                      "enum": Array [
                        "success",
                      ],
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "status",
                    "data",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Sucessful response",
          },
        },
        "security": Array [
          Object {
            "bearerAuth": Array [],
          },
        ],
        "summary": "patchTodo",
        "tags": Array [
          "todo",
        ],
      },
      "put": Object {
        "operationId": "updateTodo",
        "parameters": Array [
          Object {
            "description": "ID of the item to return",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "minimum": 0,
              "type": "integer",
            },
          },
        ],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "done": Object {
                    "type": "boolean",
                  },
                  "name": Object {
                    "maxLength": 255,
                    "minLength": 1,
                    "type": "string",
                  },
                },
                "required": Array [
                  "name",
                  "done",
                ],
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "data": Object {
                      "properties": Object {
                        "done": Object {
                          "type": "boolean",
                        },
                        "id": Object {
                          "minimum": 0,
                          "type": "integer",
                        },
                        "name": Object {
                          "maxLength": 255,
                          "minLength": 1,
                          "type": "string",
                        },
                      },
                      "required": Array [
                        "name",
                        "done",
                        "user_id",
                        "created_at",
                        "id",
                      ],
                      "type": "object",
                    },
                    "status": Object {
                      "enum": Array [
                        "success",
                      ],
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "status",
                    "data",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Sucessful response",
          },
        },
        "security": Array [
          Object {
            "bearerAuth": Array [],
          },
        ],
        "summary": "updateTodo",
        "tags": Array [
          "todo",
        ],
      },
    },
  },
  "security": Array [],
  "servers": Array [],
  "tags": Array [
    Object {
      "description": "User operations",
      "name": "user",
    },
  ],
}
`;

exports[`tr.load() loads all properties; sets collection; dereferences oas 2`] = `
Object {
  "createTodo": Object {
    "method": "post",
    "operation": Object {
      "operationId": "createTodo",
      "parameters": Array [],
      "requestBody": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "done": Object {
                  "type": "boolean",
                },
                "name": Object {
                  "maxLength": 255,
                  "minLength": 1,
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "done",
              ],
              "type": "object",
            },
          },
        },
        "required": true,
      },
      "responses": Object {
        "200": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "data": Object {
                    "properties": Object {
                      "done": Object {
                        "type": "boolean",
                      },
                      "id": Object {
                        "minimum": 0,
                        "type": "integer",
                      },
                      "name": Object {
                        "maxLength": 255,
                        "minLength": 1,
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "name",
                      "done",
                      "user_id",
                      "created_at",
                      "id",
                    ],
                    "type": "object",
                  },
                  "status": Object {
                    "enum": Array [
                      "success",
                    ],
                    "type": "string",
                  },
                },
                "required": Array [
                  "status",
                  "data",
                ],
                "type": "object",
              },
            },
          },
          "description": "Sucessful response",
        },
      },
      "security": Array [
        Object {
          "bearerAuth": Array [],
        },
      ],
      "summary": "createTodo",
      "tags": Array [
        "todo",
      ],
    },
    "path": "/api/todo",
  },
  "deleteTodo": Object {
    "method": "delete",
    "operation": Object {
      "operationId": "deleteTodo",
      "parameters": Array [
        Object {
          "description": "ID of the item to return",
          "in": "path",
          "name": "id",
          "required": true,
          "schema": Object {
            "minimum": 0,
            "type": "integer",
          },
        },
      ],
      "responses": Object {
        "200": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "data": Object {
                    "enum": Array [
                      true,
                    ],
                    "type": "boolean",
                  },
                  "status": Object {
                    "enum": Array [
                      "success",
                    ],
                    "type": "string",
                  },
                },
                "required": Array [
                  "status",
                  "data",
                ],
                "type": "object",
              },
            },
          },
          "description": "Successful operation",
        },
      },
      "security": Array [
        Object {
          "bearerAuth": Array [],
        },
      ],
      "summary": "deleteTodo",
      "tags": Array [
        "todo",
      ],
    },
    "path": "/api/todo/{id}",
  },
  "listTodo": Object {
    "method": "get",
    "operation": Object {
      "operationId": "listTodo",
      "parameters": Array [],
      "responses": Object {
        "200": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "data": Object {
                    "items": Object {
                      "properties": Object {
                        "done": Object {
                          "type": "boolean",
                        },
                        "id": Object {
                          "minimum": 0,
                          "type": "integer",
                        },
                        "name": Object {
                          "maxLength": 255,
                          "minLength": 1,
                          "type": "string",
                        },
                      },
                      "required": Array [
                        "name",
                        "done",
                        "user_id",
                        "created_at",
                        "id",
                      ],
                      "type": "object",
                    },
                    "type": "array",
                  },
                  "status": Object {
                    "enum": Array [
                      "success",
                    ],
                    "type": "string",
                  },
                },
                "required": Array [
                  "status",
                  "data",
                ],
                "type": "object",
              },
            },
          },
          "description": "Sucessful response",
        },
      },
      "security": Array [
        Object {
          "bearerAuth": Array [],
        },
      ],
      "summary": "listTodo",
      "tags": Array [
        "todo",
      ],
    },
    "path": "/api/todo",
  },
  "patchTodo": Object {
    "method": "patch",
    "operation": Object {
      "operationId": "patchTodo",
      "parameters": Array [
        Object {
          "description": "ID of the item to return",
          "in": "path",
          "name": "id",
          "required": true,
          "schema": Object {
            "minimum": 0,
            "type": "integer",
          },
        },
      ],
      "requestBody": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "minProperties": 1,
              "properties": Object {
                "done": Object {
                  "type": "boolean",
                },
                "name": Object {
                  "maxLength": 255,
                  "minLength": 1,
                  "type": "string",
                },
              },
              "type": "object",
            },
          },
        },
        "required": true,
      },
      "responses": Object {
        "200": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "data": Object {
                    "properties": Object {
                      "done": Object {
                        "type": "boolean",
                      },
                      "id": Object {
                        "minimum": 0,
                        "type": "integer",
                      },
                      "name": Object {
                        "maxLength": 255,
                        "minLength": 1,
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "name",
                      "done",
                      "user_id",
                      "created_at",
                      "id",
                    ],
                    "type": "object",
                  },
                  "status": Object {
                    "enum": Array [
                      "success",
                    ],
                    "type": "string",
                  },
                },
                "required": Array [
                  "status",
                  "data",
                ],
                "type": "object",
              },
            },
          },
          "description": "Sucessful response",
        },
      },
      "security": Array [
        Object {
          "bearerAuth": Array [],
        },
      ],
      "summary": "patchTodo",
      "tags": Array [
        "todo",
      ],
    },
    "path": "/api/todo/{id}",
  },
  "showTodo": Object {
    "method": "get",
    "operation": Object {
      "operationId": "showTodo",
      "parameters": Array [
        Object {
          "description": "ID of the item to return",
          "in": "path",
          "name": "id",
          "required": true,
          "schema": Object {
            "minimum": 0,
            "type": "integer",
          },
        },
      ],
      "responses": Object {
        "200": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "data": Object {
                    "properties": Object {
                      "done": Object {
                        "type": "boolean",
                      },
                      "id": Object {
                        "minimum": 0,
                        "type": "integer",
                      },
                      "name": Object {
                        "maxLength": 255,
                        "minLength": 1,
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "name",
                      "done",
                      "user_id",
                      "created_at",
                      "id",
                    ],
                    "type": "object",
                  },
                  "status": Object {
                    "enum": Array [
                      "success",
                    ],
                    "type": "string",
                  },
                },
                "required": Array [
                  "status",
                  "data",
                ],
                "type": "object",
              },
            },
          },
          "description": "Sucessful response",
        },
      },
      "security": Array [
        Object {
          "bearerAuth": Array [],
        },
      ],
      "summary": "showTodo",
      "tags": Array [
        "todo",
      ],
    },
    "path": "/api/todo/{id}",
  },
  "updateTodo": Object {
    "method": "put",
    "operation": Object {
      "operationId": "updateTodo",
      "parameters": Array [
        Object {
          "description": "ID of the item to return",
          "in": "path",
          "name": "id",
          "required": true,
          "schema": Object {
            "minimum": 0,
            "type": "integer",
          },
        },
      ],
      "requestBody": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "done": Object {
                  "type": "boolean",
                },
                "name": Object {
                  "maxLength": 255,
                  "minLength": 1,
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "done",
              ],
              "type": "object",
            },
          },
        },
        "required": true,
      },
      "responses": Object {
        "200": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "data": Object {
                    "properties": Object {
                      "done": Object {
                        "type": "boolean",
                      },
                      "id": Object {
                        "minimum": 0,
                        "type": "integer",
                      },
                      "name": Object {
                        "maxLength": 255,
                        "minLength": 1,
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "name",
                      "done",
                      "user_id",
                      "created_at",
                      "id",
                    ],
                    "type": "object",
                  },
                  "status": Object {
                    "enum": Array [
                      "success",
                    ],
                    "type": "string",
                  },
                },
                "required": Array [
                  "status",
                  "data",
                ],
                "type": "object",
              },
            },
          },
          "description": "Sucessful response",
        },
      },
      "security": Array [
        Object {
          "bearerAuth": Array [],
        },
      ],
      "summary": "updateTodo",
      "tags": Array [
        "todo",
      ],
    },
    "path": "/api/todo/{id}",
  },
}
`;
